docs/design/diagnostics.md

```diff
@@ -0,0 +1,293 @@
+
+```protobuf
+message ValidateProviderConfig {
+    message Request {
+        DynamicValue config = 1;
+    }
+    message Response {
+        repeated Diagnostic diagnostics = 2;
+    }
+}
+```
+
+All RPCs except `StopProvider` include diagnostics support.
+
+### terraform-plugin-go
+
+The `terraform-plugin-go` library, which underpins this framework, provides the following implementation (both `tfprotov5` and `tfprotov6`) of diagnostics:
+
+```go
+package tfprotov6
+
+type Diagnostic struct {
+    Severity  DiagnosticSeverity
+    Summary   string
+    Detail    string
+    Attribute *tftypes.AttributePath
+}
+```
+
+These types include no helper methods or slice type alias.
+
+## Prior Implementations
+
+### terraform-plugin-sdk
+
+The previous framework provided a thin abstraction layer in the [`diag` package](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/diag):
+
+```go
+type Diagnostic struct {
+    Severity      Severity
+    Summary       string
+    Detail        string
+    AttributePath cty.Path
+}
+
+func (d Diagnostic) Validate() error
+
+type Diagnostics []Diagnostic
+
+func (diags Diagnostics) HasError() bool
+
+func Errorf(format string, a ...interface{}) Diagnostics
+
+func FromErr(err error) Diagnostics
+```
+
+## Goals
+
+Diagnostics in this framework should be:
+
+- Able to natively support warning and error diagnostics.
+- Implementable by provider developers.
+- Abstracted from terraform-plugin-go and convertable into those types to separate implementation concerns.
+- Ergonomic to create, read, and delete in Go code (e.g. have helper methods for common use cases).
+
+Additional consideration should be given to:
+
+- Extensibility and additional data storage (e.g. distinguishing between types of diagnostics or what generated an error).
+- The ability to deduplicate or manipulate diagnostics.
+
+## Proposals
+
+### Thin Abstraction
+
+The most trivial implementation would be creating new Go types in this framework that provide a thin abstraction layer, e.g.
+
+```go
+package diag
+
+type Diagnostic struct {
+    Severity  DiagnosticSeverity
+    Summary   string
+    Detail    string
+    Attribute *tftypes.AttributePath // See also: https://github.com/hashicorp/terraform-plugin-framework/issues/81
+}
+
+type Diagnostics []Diagnostic
+
+func (d Diagnostics) Append(in ...Diagnostic)
+func (d Diagnostics) HasError() bool
+func (d Diagnostics) toTfprotov6Diagnostics() []*tfprotov6.Diagnostic
+```
+
+This solves many requirements, allowing the framework to own the abstraction and provide ergonomic methods. However, this implementation is tightly coupled to the underlying implementation and does not introduce protections against backwards compatibility issues. Since the data fields are strongly associated with the type, implementors inside and outside the framework are bound to the initial implementation details. If those details are not exported and instead hidden behind a constructor function, e.g.
```

Comment:
Can you expand on _why_ it would be difficult to change? Elsewhere (e.g., request/response pattern) we've argued that using structs _helps_ backwards compatibility. What sets this apart?

Reply:
Removed 👍